File: packages/core/src/attacks/hit.ts
export type HitRateParams = {
  accuracyTerm: number
  evasionTerm: number
  moraleModifier: number

  criticalRateMultiplier: number
  hitRateBonus?: number
  criticalRateBonus?: number
}

export type HitRateDetail = HitRateParams & {
  hitRate: number
  criticalRate: number
  normalRate: number
}

/**
 * @returns 10~96
 */
const calcHitRateBasis = ({ accuracyTerm, evasionTerm, moraleModifier }: HitRateParams) => {
  const value = (accuracyTerm - evasionTerm) * moraleModifier

  if (value < 10) {
    return 10
  }

  if (value > 96) {
    return 96
  }

  return value
}

export const calcHitRate = (params: HitRateParams): HitRateDetail => {
  const { criticalRateMultiplier, hitRateBonus = 0, criticalRateBonus = 0 } = params
  const hitRateBasis = calcHitRateBasis(params)

  const hitPercent = Math.floor(hitRateBasis + 1 + hitRateBonus)
  const criticalPercent = Math.floor(Math.sqrt(hitRateBasis) * criticalRateMultiplier + 1 + criticalRateBonus * 100)

  const hitRate = Math.min(hitPercent / 100, 1)
  const criticalRate = Math.min(criticalPercent / 100, 1)
  const normalRate = hitRate - criticalRate

  return { ...params, hitRate, criticalRate, normalRate }
}

type HitStatus = "Miss" | "Normal" | "Critical"

export const getHitStatus = (rate: Pick<HitRateDetail, "criticalRate" | "hitRate">): HitStatus => {
  const randomNum = Math.random()

  if (randomNum < rate.criticalRate) return "Critical"
  if (randomNum < rate.hitRate) return "Normal"
  return "Miss"
}

File: packages/core/src/attack/hit.rs
use serde::Serialize;
use ts_rs::TS;

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum HitType {
    Miss,
    Normal,
    Critical,
}

#[derive(Debug, Default, Serialize, TS)]
pub struct HitRateParams {
    pub accuracy_term: f64,
    pub evasion_term: f64,
    pub morale_mod: f64,
    pub critical_rate_multiplier: f64,
    pub critical_percent_bonus: f64,
    pub hit_percent_bonus: f64,
}

#[derive(Debug, Default, Serialize, TS)]
pub struct HitRate {
    pub total: f64,
    pub normal: f64,
    pub critical: f64,
}

impl HitRateParams {
    fn calc_basis(&self) -> f64 {
        let value = (self.accuracy_term - self.evasion_term) * self.morale_mod;
        value.clamp(10.0, 96.0)
    }

    pub fn calc(&self) -> HitRate {
        let basis = self.calc_basis();

        let critical_percent =
            (basis.sqrt() * self.critical_rate_multiplier + 1.0 + self.critical_percent_bonus)
                .floor();

        let hit_percent = (basis + 1.0 + self.hit_percent_bonus)
            .floor()
            .max(critical_percent);

        let total = (hit_percent / 100.0).min(1.0);
        let critical = (critical_percent / 100.0).min(1.0);
        let normal = total - critical;

        HitRate {
            total,
            normal,
            critical,
        }
    }
}

#[cfg(test)]
mod test {
    use super::HitRateParams;

    #[test]
    fn test_hit_rate() {
        let min = HitRateParams {
            accuracy_term: 0.0,
            evasion_term: 0.0,
            morale_mod: 1.0,
            critical_rate_multiplier: 1.0,
            critical_percent_bonus: 0.0,
            hit_percent_bonus: 0.0,
        };

        let max = HitRateParams {
            accuracy_term: 100.0,
            hit_percent_bonus: 10.0,
            ..min
        };

        assert_eq!(min.calc_basis(), 10.0);
        assert_eq!(max.calc_basis(), 96.0);
        assert_eq!(max.calc().total, 1.0);

        let hit_rate = HitRateParams {
            accuracy_term: 60.0,
            evasion_term: 10.0,
            morale_mod: 1.2,
            critical_rate_multiplier: 1.3,
            critical_percent_bonus: 12.0,
            hit_percent_bonus: 9.0,
        }
        .calc();

        assert_eq!(hit_rate.total, 0.7);
        assert_eq!(hit_rate.critical, 0.23);
        assert_eq!(hit_rate.normal, 0.7 - 0.23)
    }
}
